commons-app/src/main/java/cz/cuni/mff/xrg/odcs/commons/app/auth/DefaultPermissionEvaluator.java

```diff
@@ -60,48 +93,88 @@ public boolean hasPermission(Authentication auth, Object target, Object perm) {
-                    if (ShareType.PUBLIC_RW.equals(sTarget.getShareType())) {
+            if (foundPermission != null) {
+                if (foundPermission.isRwOnly() && !ShareType.PUBLIC_RW.equals(sTarget.getShareType())) {
+                    return false;
+                } else {
+                    //only owner can delete
+                    if(!"pipeline.delete".equals(requestedPerm)){
                         return true;
+                    } else {
+                        return false;
                     }
-                    break;
-                case "delete":
-                    // refuse delete, only for owner or admin
-                    break;
+                }
             }
         }
 
         // Pipeline execution actions are always viewable if
         // pipeline itself is viewable
         if (target instanceof PipelineExecution) {
-            switch (perm.toString()) {
-                case "view":
-                    Pipeline pipe = ((PipelineExecution) target).getPipeline();
-                    boolean viewPipe = hasPermission(pipe, perm);
-                    if (viewPipe) {
-                        // user has permission to view pipeline
-                        // for this execution -> allow to see execution as well
-                        return true;
-                    }
-                    break;
+            Pipeline pipe = ((PipelineExecution) target).getPipeline();
+            boolean viewPipe = hasPermission(pipe, requestedPerm);
+            if (viewPipe) {
+                // user has permission to view pipeline
+                // for this execution -> allow to see execution as well
+                return true;
+            }
+        }
+
+        // in other cases be restrictive
+        LOG.debug("Method hasPermission refused access for object <{}> and permission <{}>.",
+                target, requestedPerm);
+        return false;
+    }
+
+    private boolean hasPermissionOrganization(Authentication auth, Object target, String requestedPerm, Permission foundPermission) {
+
+        User user = (User) auth.getPrincipal();
+
+        if (target instanceof OrganizationSharedEntity) {
+            OrganizationSharedEntity oTarget = (OrganizationSharedEntity) target;
+            Organization organization = oTarget.getOrganization();
+
+            if (organization != null) {
+                if (user.getOrganization().getName().equals(organization.getName())) {
+                    return true;
+                }
+            }
+        }
+
+        // entity owner is almighty
+        if (target instanceof OwnedEntity) {
+            OwnedEntity oTarget = (OwnedEntity) target;
+            User owner = oTarget.getOwner();
+            if (owner != null && auth.getName().equals(owner.getUsername())) {
+                return true;
+            }
+        }
+
+        if (target instanceof DPUTemplateRecord) {
+            DPUTemplateRecord dpuTarget = (DPUTemplateRecord) target;
```

Comment:
1) Please explain what this "if" is doing. 
2) Why is this if not included in hasPermissionUser?